rcli/typing.py

```diff
@@ -0,0 +1,671 @@
+    """A metaclass to turn a class into a singleton."""
+
+    __instance__ = None  # type: type
+
+    def __call__(cls, *args, **kwargs):
+        # type: (*Any, **Any) -> type
+        """Instantiate the class only once."""
+        if not cls.__instance__:
+            cls.__instance__ = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls.__instance__
+
+
+class _Uninstantiable(type):
+    """A metaclass that disallows instantiation."""
+
+    def __call__(cls, *args, **kwargs):
+        # type: (*Any, **Any) -> None
+        """Do not allow the class to be instantiated."""
+        raise TypeError('Type {} cannot be instantiated.'.format(cls.__name__))
+
+
+class _ClsReprMeta(type):
+    """A metaclass that returns a custom type repr if defined."""
+
+    __class_repr__ = None  # type: Optional[str]
+
+    def __repr__(cls):
+        # type: () -> str
+        """Return a custom string for the subclass if defined."""
+        if cls.__class_repr__:
+            return cls.__class_repr__
+        return super(cls.__class__, cls).__repr__()
+
+
+class _BoundedMeta(_Uninstantiable):
+    """A metaclass that adds slicing to a class that creates new classes."""
+
+    def __getitem__(cls, args):
+        # type: (Union[Tuple[_T, Any], Tuple[_T, Any, Callable]]) -> type
+        """Create a new subclass of a type bounded by the arguments.
+
+        If a callable is passed as the third argument of the slice, it will be
+        used as the comparison function for the boundaries.
+
+        Args:
+            args: A tuple with two or three parameters: a type, a slice
+                representing the minimum and maximum lengths allowed for values
+                of that type and, optionally, a function to use on values
+                before comparing against the bounds.
+        """
+        type_, bound, keyfunc = cls._get_args(args)
+        keyfunc_name = cls._get_fullname(keyfunc)
+        identity = cls._identity
+        try:
+            class BaseClass(type_):
+                """Use a subclass of type_ if it is subclassable."""
+        except TypeError:
+            BaseClass = object
+
+        @six.add_metaclass(_ClsReprMeta)
+        class _BoundedSubclass(BaseClass):  # type: ignore
+            """A subclass of a type bounded by a slice."""
+
+            def __new__(cls, __value, *args, **kwargs):
+                # type: (Any, *Any, **Any) -> _T
+                """Return __value cast to _T.
+
+                Any additional arguments are passed as-is to the constructor.
+
+                Args:
+                    __value: A value that can be converted to type _T.
+                    args: Any additional positional arguments passed to the
+                        constructor.
+                    kwargs: Any additional keyword arguments passed to the
+                        constructor.
+                """
+                try:
+                    ret = type_(__value, *args, **kwargs)
```

Comment:
I vote that the try: except: assigns "BaseClass = type_".  (about line 138)

Then on line 160 on the RHS of "=" instantiate `BaseClass`